x-pack/osquerybeat/magefile.go

```diff
@@ -77,10 +82,103 @@ func Clean() error {
 	return devtools.Clean(paths)
 }
 
+func extractFromMSI() error {
+	if os.Getenv("GOOS") != "windows" {
+		return nil
+	}
+
+	ctx := context.Background()
+
+	execCommand := func(name string, args ...string) error {
+		ps := strings.Join(append([]string{name}, args...), " ")
+		fmt.Println(ps)
+		output, err := command.Execute(ctx, name, args...)
+		if err != nil {
+			fmt.Println(ps, ", failed: ", err)
+			return err
+		}
+		fmt.Print(output)
+		return err
+	}
+
+	// Install msitools
+	err := execCommand("apt", "update")
+	if err != nil {
+		return err
+	}
+
+	err = execCommand("apt", "install", "-y", "msitools")
+	if err != nil {
+		return err
+	}
+
+	osArchs := osquerybeat.OSArchs(devtools.Platforms)
+
+	for _, osarch := range osArchs {
+		if osarch.OS != "windows" {
+			continue
+		}
+		spec, err := distro.GetSpec(osarch)
+		if err != nil {
+			if errors.Is(err, distro.ErrUnsupportedOS) {
+				continue
+			} else {
+				return err
+			}
+		}
+		dip := distro.GetDataInstallDir(osarch)
+		msiFile := spec.DistroFilepath(dip)
+
+		// MSI extract
+		err = execCommand("msiextract", "--directory", dip, msiFile)
+		if err != nil {
+			return err
+		}
+
+		fmt.Println("copy osqueryd.exe from MSI")
+		dp := distro.OsquerydPathForOS(osarch.OS, dip)
+		err = devtools.Copy(filepath.Join(dip, "osquery", "osqueryd", "osqueryd.exe"), dp)
+		if err != nil {
+			fmt.Println("copy osqueryd.exe from MSI failed: ", err)
+			return err
+		}
+		// Chmod set to the same as other executables in the final package
+		if err = os.Chmod(dp, 0755); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
 // GolangCrossBuild build the Beat binary inside of the golang-builder.
 // Do not use directly, use crossBuild instead.
 func GolangCrossBuild() error {
-	return devtools.GolangCrossBuild(devtools.DefaultGolangCrossBuildArgs())
+	// This is to fix a defect in the field where msiexec fails to extract the osqueryd.exe
+	// from bundled osquery.msi, with error code 1603
+	// https://docs.microsoft.com/en-us/troubleshoot/windows-server/application-management/msi-installation-error-1603
+	// SDH: https://github.com/elastic/sdh-beats/issues/1575
+	// Currently we can't reproduce this is issue, but here we can eliminate the need for calling msiexec
+	// if extract the osqueryd.exe binary during the build.
+	//
+	// The builder docker images are Debian so we need to install msitools for
+	// linux in order to extract the osqueryd.exe from MSI during build process.	// Install MSI tools in order to extract file from MSI
+	// Ideally we would want these to be a part of the build docker image,
+	// but doing this here for now due to limited time before 7.16.2
```

Comment:
Is there an item to track this once we are not pressed for time?

Reply:
not at the moment